basic/product-marketplace/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+        buyer_nft_id_by_product_id : HashMap<u128, u128>,
+        admin_badge: Address,
+        seller_product_minter_badge_vault : Vault,
+        buyer_minter_badge_vault : Vault,
+        seller_permanent_badge_vault : Vault,
+        seller_product_badge_def : ResourceDef,
+        buyer_badge_def :  ResourceDef,
+        seller_permanent_badge_def : ResourceDef,
+        vault_by_seller : HashMap<u128,Vault>,
+        sell_fees : Decimal,
+        buy_fees : Decimal,
+        token_type : Address,
+        payment_by_buyer_nft_id : HashMap<u128,Vault>, 
+    }
+
+    impl ProductMarketPlace {
+        // Implement the functions and methods which will manage those resources and data
+        
+        // This is a function, and can be called directly on the blueprint once deployed
+        pub fn new(token_type : Address, 
+                   sell_fees : Decimal, 
+                   buy_fees : Decimal) -> (Component, Bucket) {
+            
+                    
+            let admin_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+            .initial_supply_fungible(1);
+
+            let seller_product_minter_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+            .metadata("name", "seller minter product badge")
+            .initial_supply_fungible(1);
+
+            let seller_permanent_minter_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+            .metadata("name", "seller minter permanent badge")
+            .initial_supply_fungible(1);
+
+            let seller_product_badge_def = ResourceBuilder::new_non_fungible()
+            .metadata("name", "seller badge")
+            .metadata("description", "this badge give to seller rigth to get postal stamp to send product to buyer.")
+            .flags(MINTABLE | INDIVIDUAL_METADATA_MUTABLE | BURNABLE)
+            .badge(seller_product_minter_badge.resource_address(), MAY_MINT | MAY_CHANGE_INDIVIDUAL_METADATA | MAY_BURN)
+            .no_initial_supply(); 
+            
+            let seller_permanent_badge_def = ResourceBuilder::new_non_fungible()
+            .metadata("name", "permanent seller badge")
+            .metadata("description", "this badge give to seller rigth to collect money")
+            .flags(MINTABLE | INDIVIDUAL_METADATA_MUTABLE)
+            .badge(seller_permanent_minter_badge.resource_address(), MAY_MINT | MAY_CHANGE_INDIVIDUAL_METADATA)
+            .no_initial_supply(); 
+            
+            let buyer_minter_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+            .metadata("name", "buyer minter badge")
+            .initial_supply_fungible(1);
+
+
+            let buyer_badge_def = ResourceBuilder::new_non_fungible()
+            .metadata("name", "buyer badge")
+            .flags(MINTABLE | INDIVIDUAL_METADATA_MUTABLE | BURNABLE)
+            .badge(buyer_minter_badge.resource_address(), MAY_MINT | MAY_CHANGE_INDIVIDUAL_METADATA | MAY_BURN)
```

Comment:
Is there a particular reason that you use three different badges to mint the three tokens ? I think it would simplify the code to only use one badge that is used for all resources

Reply:
Come to think of it, one badge is enough. I will fix it.

Reply:
Hi clement, I finally kept the badge allowing sellers to collect their recipe to distinguish it from the one associated with the product. The first cannot be burnt but the second yes.